File: Compass/Agents/GithubCopilot.cs
namespace Compass.Agents;

public class GithubCopilot : Agent
{
    public override string Name => "GitHub Copilot";

    public override Task EnsureLogin()
    {
        // do nothing for now, since there is no non-interactive way to check this.
        return Task.CompletedTask;
    }

    public override async Task<AgentOutput> Execute(string prompt, string model, string workingDirectory)
    {
        var processOutput = await ProcessUtils.Run(
            workingDirectory,
            "copilot",
            $"--silent --no-color --model {model.EscapeArg()} --allow-all-tools --allow-all-paths --add-dir {workingDirectory.EscapeArg()} -p {prompt.EscapeArg()}");

        Logger.Log($"Collecting git diff after agent execution", Logger.LogLevel.Verbose);
        
        var diff = await ProcessUtils.Git(workingDirectory, "--no-pager diff");

        return new AgentOutput()
        {
            StdOut = processOutput.StdOut,
            StdErr = processOutput.StdErr,
            GitDiff = diff.StdOut
        };
    }
}

File: compass/agents/githubCopilot.ts
import type { AgentOutput } from "../models";
import { escapeArg, logger, run } from "../utils";
import { Agent } from "./agent";

export class GitHubCopilot extends Agent {
    constructor() {
        super("GitHub Copilot");
    }

    override async execute(prompt: string, model: string, workingDirectory: string): Promise<AgentOutput> {
        logger.info(`Executing GitHub Copilot with model ${model} on prompt ${prompt}`);
        
        const processOutput = await run(
            workingDirectory,
            "copilot",
            "--silent", 
            "--no-color", 
            "--model", model, 
            "--allow-all-tools", 
            "--allow-all-paths", 
            "--add-dir", workingDirectory, 
            "-p", prompt);

        logger.trace("Collecting git diff after agent execution");
        
        const diff = await run(workingDirectory, "git", "--no-pager", "diff");

        return {
            stdOut: processOutput.stdOut,
            stdErr: processOutput.stdErr,
            exitCode: processOutput.exitCode,
            gitDiff: diff.stdOut
        };
    }

    override init(): Promise<void> {
        return Promise.resolve();    
    }
    
}